### DISCUSSION TITLE
Softmax and binary classification problem in MoleculeNet

### DISCUSSION THREAD

**Discussion by jiaruHithub** [pyg-team/pytorch_geometric#5597] [Q&A]
In ```MoleculeNet```, there is many binary classfication problem datasets. In general, ```BCE loss``` should be used during training on the datasets of ```MoleculeNet```. But, I generated a generic representation ```g_rep``` for each class of data in a dataset, When a graph is represented by GNN, I want the representation to match the generic vector ```g_rep```, and the class corresponding to the vector ```g_rep``` with the highest matching possibility is the label of the graph.

Thus I use softmax to output the matching possibility between current graph representation and every vector ```g_rep```, then I use ```F.nll_loss``` to compute the loss. And the target is ```label```, Because the order of ```g_rep``` for each class is the same as the order of label in one-hot label.

But surprisingly, the accuracy rate of the network is only 50%, and it stays the same no matter how hard it is trained. Is there something wrong with me? Help. Thank you very much.

**Comment by rusty1s** [comment]
I might be missing something, but if you want to train your GNN to output the representation of `g_rep`, wouldn't you need to train with MSE loss? Then, during inference you would pick the label of the closest embedding.

**Reply by jiaruHithub** [reply]
oh,sorry, my Grammer error. The ```g_rep``` is generated by a pretrained GNN1 use many graphs in a same class. Current graph representation ```x_rep``` is generated another pretrained GNN2. I use the matching probability as the result. For example, a dataset has two class 0 and 1, thus I generate ```g_rep0``` and ```g_rep1```, if current ```x_rep``` is matching with ```g_rep1```, the label of ```x_rep``` is 1. so I use softmax to output the matching probability, and use null_loss

**Reply by rusty1s** [reply]
But how do you match in the first place? If you are simply using a discrete matching operation here, there will be no gradients flying back to the GNN.

**Reply by jiaruHithub** [reply]
I use ```nn.linear``` for matching, which is feed a concatenated vectors, ```torch.concat(x_pre, g_rep0)``` and ```torch.concat(x_pre, g_rep1)```. I think gradient is existing, right?

**Reply by rusty1s** [reply]
Yes, in that case gradients should exist. Can you show a minimal example of your loss computation? Might help to find the issue.

**Reply by jiaruHithub** [reply]
Of course, my train module is:
```
model.train()

for step, batch in enumerate(tqdm(loader, desc="Iteration")):
    batch = batch.to(device)
 
    t_batch = t_batch.to(device)#this is g_rep vector #t_batch contains two general graph, which
    temp_rep = gnn_p(t_batch.x, t_batch.edge_index, t_batch.edge_attr, t_batch.batch)
    graph_rep = gnn_g(batch.x, batch.edge_index, batch.edge_attr, batch.batch)
    repeat_num = graph_rep.shape[0]
    temp_rep = torch.repeat_interleave(temp_rep, repeat_num, dim = 0)
    graph_rep = torch.repeat_interleave(graph_rep, 2, dim = 0)
    pred = pro_head(torch.cat((graph_rep, temp_rep), 1)).view(-1, 2)
    pred = F.softmax(pred, dim = 1)

    y = batch.y.view(repeat_num, -1).to(torch.float64)
 
    is_valid = y**2 > 0
  
    loss_mat = F.nll_loss(pred, ((y+1)/2).squeeze().long())
    #loss matrix after removing null target
    loss_mat = torch.where(is_valid, loss_mat, torch.zeros(loss_mat.shape).to(loss_mat.device).to(loss_mat.dtype))
        
    optimizer.zero_grad()
    loss = torch.sum(loss_mat)/torch.sum(is_valid)
    loss.backward()

    optimizer.step()
```
eval module is
```
model.eval()
y_true = []
y_scores = []

for step, batch in enumerate(loader):
    batch = batch.to(device)
    t_batch = t_batch.to(device)

    with torch.no_grad():
        temp_rep = gnn_p(t_batch.x, t_batch.edge_index, t_batch.edge_attr, t_batch.batch)
        graph_rep = gnn_g(batch.x, batch.edge_index, batch.edge_attr, batch.batch)
        
        repeat_num = graph_rep.shape[0]
        temp_rep = torch.repeat_interleave(temp_rep, repeat_num, dim = 0)
        graph_rep = torch.repeat_interleave(graph_rep, 2, dim = 0)
        pred = pro_head(torch.cat((graph_rep, temp_rep), 1)).view(-1, 2)
        pred = F.softmax(pred, dim = 1)
    
    y_true.append(batch.y.view(repeat_num, -1))
    
    y_scores.append(pred)

y_true = torch.cat(y_true, dim = 0).cpu().numpy()
y_scores = torch.cat(y_scores, dim = 0).cpu().numpy()

roc_list = []
for i in range(y_true.shape[1]):
    #AUC is only defined when there is at least one positive data.
    if np.sum(y_true[:,i] == 1) > 0 and np.sum(y_true[:,i] == -1) > 0:
        is_valid = y_true[:,i]**2 > 0
        roc_list.append(roc_auc_score((y_true[is_valid,i] + 1)/2, y_scores[is_valid,i]))

**Reply by jiaruHithub** [reply]
```gnn_p``` and ```gnn_g``` is initialized from same pretrained gnn model.

**Reply by rusty1s** [reply]
Is your `pro_head` model trained?

**Reply by jiaruHithub** [reply]
No, it is a random initiatived full-connection layer to output the matching probability.

**Reply by jiaruHithub** [reply]
and it is trained with loss, not pretrained model

**Reply by rusty1s** [reply]
Can you clarify? If `pro_head` is random initialized, then this is equal to a random guess.

**Reply by jiaruHithub** [reply]
I m sure it's random initalized. Could you please tell me why it equal to random guess? Why this full-connected layer must be pretrained?

**Reply by jiaruHithub** [reply]
Doesn't that '''pro_head''' train with loss？
Another strange phenomenon if I comment ```pred = F.softmax(pred, dim = 1)``` , the networ will be trained, but acc will stay around 60%

**Reply by rusty1s** [reply]
Yes, it should train with loss (I thought you just run this during inference as shown in your example). One reason that the softmax might fail is that you apply a RELU at the end of `pro_head`. Is that the case?

**Reply by jiaruHithub** [reply]
if I use softmax, the acc will always stay 50.000%, like random guess. And I had not use relu function in the end of ```pro_head```. You mean I can try relu ?

**Reply by rusty1s** [reply]
Mh, no you shouldn't use relu at the end (I just thought it might explain this behavior). Please note that if you are using `nll_loss`, you have to use `log_softmax` instead.
